northd/multicast.dl

```diff
@@ -117,6 +135,61 @@ SwitchMcastFloodRelayPorts(switch, set_empty()) :-
         }
     ).
 
+/* Mapping between Switch and the set of port uuids on which to
+ * flood IP multicast statically.
+ */
+relation SwitchMcastFloodPorts(sw: Ref<Switch>, ports: Set<uuid>)
```

Comment:
There are several instances of using `Ref<Switch>` as a key (i.e., a field used to match records in this relation to other relation) in this PR. This can be expensive, and, as far as I can tell, is unnecessary. `Switch` is a complex datatype with many fields.  Even though it is wrapped in a `Ref`, most operations like comparison and hashing are still expensive for it (`Ref<>` saves memory and speeds up equality checks (`==`), but not any of the other operations on a type). As a result, indexing the relation by this field is more expensive than indexing by, say, uuid.  It is hard to tell how bad this is in practice without careful benchmarking, but I think that it is generally a good practice to avoid this.

One alternative is to store switch UUID only instead of `Ref<Switch>` in the relation:

```
relation SwitchMcastFloodPorts(sw: uuid, ports: Set<uuid>)
```

then instead of

```
for (sw in &Switch(.dpname = dpname, .mcast_cfg = &mcast_cfg)
  for (IgmpSwitchMulticastGroup(.address = address, .switch = &sw)) {
    for (SwitchMcastFloodRelayPorts(&sw, relay_ports)) {
        for (SwitchMcastFloodPorts(&sw, flood_ports)) { 
```

you can write

```
for (sw in &Switch(.ls = Logical_Switch{._uuid=sw_uuid}, .dpname = dpname, .mcast_cfg = &mcast_cfg)
  for (IgmpSwitchMulticastGroup(.address = address, .switch = sw_uuid)) {
    for (SwitchMcastFloodRelayPorts(sw_uuid, relay_ports)) {
        for (SwitchMcastFloodPorts(sw_uuid, flood_ports)) { 
```

Reply:
Oh, my bad, I completely misunderstood how `Ref` works. I assumed it's just the ddlog equivalent of C pointer (or reference in other languages).

But the main reason why I chose `Ref<Switch>` and `Ref<Router>` is the type safety we get from it. As far as I see we lose all that if we change to `uuid` as the compiler has no way to distinguish between the type of the records we're referencing.

However, if this turns out to be a bottle neck, maybe there's a way to optimize references in DDlog or maybe even add a new type?

What do you think @ryzhyk ?

Thanks,
Dumitru.

Reply:
Your understanding is correct as far as the physical representation of `Ref`goes -- it's just a reference counted pointer.  However, behavior-wise, `Ref<T>` forwards all operations, including comparison and hashing to `T`. The alternative semantics would not play well with the differential dataflow engine (or at least it would take some effort to make sure that it does).

I agree with the type safety argument and that it is a good idea to keep your current implementation until it proves to be a bottleneck, in which case we can always find a different way to enforce type safety.

Reply:
@ryzhyk Sounds good, thanks.